hadoop-hdds/server-scm/src/main/java/org/apache/hadoop/hdds/scm/server/SCMBlockProtocolServer.java

```diff
@@ -198,6 +201,13 @@ public List<AllocatedBlock> allocateBlock(
             .allocateBlock(size, replicationConfig, owner, excludeList);
         if (block != null) {
           blocks.add(block);
+          // Sort the datanodes if client machine is specified
+          if (StringUtils.isNotEmpty(clientMachine)) {
+            List<String> uuidList = HddsUtils.toNodeUuid(
+                block.getPipeline().getNodes());
+            block.getPipeline().setNodesInOrder(
+                sortDatanodes(uuidList, clientMachine));
```

Comment:
Sort the datanode directly: 
```java
@@ -200,6 +203,14 @@ public List<AllocatedBlock> allocateBlock(
             .allocateBlock(size, replicationConfig, owner, excludeList);
         if (block != null) {
           blocks.add(block);
+          // Sort the datanodes if client machine is specified
+          if (StringUtils.isNotEmpty(clientMachine)) {
+            final Node client = getClientNode(clientMachine);
+            final List<DatanodeDetails> nodes = block.getPipeline().getNodes();
+            final List<DatanodeDetails> sorted = scm.getClusterMap()
+                .sortByDistanceCost(client, nodes, nodes.size());
+            block.getPipeline().setNodesInOrder(sorted);
+          }
         }
       }
 
@@ -347,11 +358,7 @@ public List<DatanodeDetails> sortDatanodes(List<String> nodes,
     auditMap.put("nodes", String.valueOf(nodes));
     try {
       NodeManager nodeManager = scm.getScmNodeManager();
-      Node client = getDatanode(clientMachine);
-      // not datanode
-      if (client == null) {
-        client = getOtherNode(clientMachine);
-      }
+      final Node client = getClientNode(clientMachine);
       List<DatanodeDetails> nodeList = new ArrayList<>();
       nodes.forEach(uuid -> {
         DatanodeDetails node = nodeManager.getNodeByUuid(uuid);
@@ -376,10 +383,10 @@ public List<DatanodeDetails> sortDatanodes(List<String> nodes,
     }
   }
 
-  private Node getDatanode(String clientMachine) {
+  private Node getClientNode(String clientMachine) {
     List<DatanodeDetails> datanodes = scm.getScmNodeManager()
         .getNodesByAddress(clientMachine);
-    return !datanodes.isEmpty() ? datanodes.get(0) : null;
+    return !datanodes.isEmpty() ? datanodes.get(0) : getOtherNode(clientMachine);
   }
 
   private Node getOtherNode(String clientMachine) {
```

Reply:
Thank you for the review. I have updated it accordingly.

I also added a server-side test in `TestSCMBlockProtocolServer` to test allocate block when client machine is specified.